### ISSUE TITLE
Main OS gets stuck in data migration

### ISSUE THREAD

**Issue by VisualEhrmanntraut** [ChefKissInc/QEMUAppleSilicon#26]
Causes:
- Lack of GPU
- Lack of Baseband (maybe also BT/WiFi)
- ~~Lack of Multitouch~~ implemented

**Comment by waterdragon78** [comment]
What’s the status on springboard and OOB setup?

**Comment by VisualEhrmanntraut** [comment]
@waterdragon78 One of the requirements to get to the UI is getting past the data migration step. The data migration step is ran even if the device has been newly restored, and it installs things like apps (Calculator, etc).

We have attempted to patch all the blocking requirements out and the UI does show up and if we bypass the activation we also get to the Home Screen. Below is how that looks like:
![2024-07-21 00 43 30](https://github.com/user-attachments/assets/4edf0d57-e46b-440c-8d5d-2d1987ac3721)
![2024-07-21 00 43 33](https://github.com/user-attachments/assets/a0531dc2-c09c-4587-a1a0-2d76096ebd8e)
![2024-07-21 00 43 36](https://github.com/user-attachments/assets/2d57248f-dade-4b9b-beaa-1e68a5748480)
![2024-07-21 00 43 37](https://github.com/user-attachments/assets/fb621ab9-e550-49c6-878e-3558e26e89e6)

**Comment by aellul27** [comment]
> Need to further implement the simulated SEP.

Is this why the ios install halts at entering copy_sep?

**Comment by VisualEhrmanntraut** [comment]
No, you did something wrong.

**Comment by Anonymous941** [comment]
> @waterdragon78 One of the requirements to get to the UI is getting past the data migration step. The data migration step is ran even if the device has been newly restored, and it installs things like apps (Calculator, etc).
> 
> We have attempted to patch all the blocking requirements out and the UI does show up and if we bypass the activation we also get to the Home Screen. Below is how that looks like:

Wow, that's amazing that you're this close - I never thought this would happen...

**Comment by KneesDev** [comment]
Any guides here, like https://github.com/TrungNguyen1909/qemu-t8030/wiki/Bringing-up-the-emulator?

**Comment by VisualEhrmanntraut** [comment]
No, sorry. This is just the upstream commits plus some patches to some files, controlled using a VNC daemon inserted into the system. It's not user-intuitive at all. We might make a blog post about this, but I'd recommend you wait until we have finished implementing enough components so patches aren't required.

**Comment by KneesDev** [comment]
Ah okay, thanks!

**Comment by GradedWarrior** [comment]
> @waterdragon78 One of the requirements to get to the UI is getting past the data migration step. The data migration step is ran even if the device has been newly restored, and it installs things like apps (Calculator, etc).
> 


Could you possibly make a docker container with everything setup?

**Comment by VisualEhrmanntraut** [comment]
> > @waterdragon78 One of the requirements to get to the UI is getting past the data migration step. The data migration step is ran even if the device has been newly restored, and it installs things like apps (Calculator, etc).
> 
> Could you possibly make a docker image with it setup?

No.

**Comment by VisualEhrmanntraut** [comment]
Here's some videos and screenshots
<details>
<summary>click to expand</summary>

https://github.com/user-attachments/assets/b4208338-920e-4e82-9a02-0423634382c4


https://github.com/user-attachments/assets/a60eb5b8-313e-43af-b0eb-eac0c169c1e5

![LockScreen_FixedAspectRatio](https://github.com/user-attachments/assets/96351274-0ca4-4e02-a641-7f1d21124eea)
![HomeScreen_FixedAspectRatio](https://github.com/user-attachments/assets/385e1602-039a-4cae-9492-37b7da557263)
</details>

**Comment by VisualEhrmanntraut** [comment]
By the way, this fork is going to become obsolete; I am in the process of writing my own emulator from scratch in Rust. This will also allow for graphics acceleration via AGX emulation and HiDPI support and faster development times with less bugs.
Once I finish on-going work with the ChefKiss website (https://chefkissinc.github.io) I am going to make a technical post in it with patches to get to the UI so you all can play around while we make a better more user-friendly version.

**Comment by aspauldingcode** [comment]
> By the way, this fork is going to become obsolete; I am in the process of writing my own emulator from scratch in Rust. This will also allow for graphics acceleration via AGX emulation and HiDPI support and faster development times with less bugs.

Holy sh*t dude you're awesome. 
I'm looking forward to this in the future. 

P.S. thanks for NootedRed. 
Awesome projects!

**Comment by Anonymous941** [comment]
Wow, this is so exciting.  I've been waiting for years for an iOS emulator to be made that's not some proprietary cloud service or a scam

**Comment by GradedWarrior** [comment]
> Here's some videos and screenshots

So fascinating! In the first video, the flashlight and camera shortcut are cut off and the icon grid is overlapping and the home bar is still at the proper location then in the later images, it seems to be all fixed, what was the cause of the former shorter screen?

**Comment by VisualEhrmanntraut** [comment]
> > Here's some videos and screenshots
> 
> So fascinating! In the first video, the flashlight and camera shortcut are cut off and the icon grid is overlapping and the home bar is still at the proper location then in the later images, it seems to be all fixed, what was the cause of the former shorter screen?

Just the aspect ratio of the resolution previously used was not expected. Seems to not be very adaptive to different aspect ratios.

**Comment by GradedWarrior** [comment]
Would your emulator allow an easy way to experiment with the aspect ratio?

**Comment by VisualEhrmanntraut** [comment]
> Would your emulator allow an easy way to experiment with the aspect ratio?

Well, obviously this issue is with iOS and not with the emulator. If the resolution is kept native the UI should look fine. I'm not sure how exactly I'm going to handle resolution, but I think it would be fun to be able to mess with the aspect ratio and cause iOS to look funky. I will have to warn you however that you can't change the resolution of the internal display while iOS is running because it reads it only once.

**Comment by GradedWarrior** [comment]
> I'm not sure how exactly I'm going to handle resolution, but I think it would be fun to be able to mess with the aspect ratio and cause iOS to look funky.

Yes, I'm interested in that.

**Comment by nick-botticelli** [comment]
> @waterdragon78 One of the requirements to get to the UI is getting past the data migration step. The data migration step is ran even if the device has been newly restored, and it installs things like apps (Calculator, etc).
> 
> We have attempted to patch all the blocking requirements out and the UI does show up and if we bypass the activation we also get to the Home Screen. Below is how that looks like:
> 
> Screenshots

Hey, I'm super curious what patches you made to get the UI to show up! I've been trying to get past the same Data Migration step on iOS 15.

**Comment by VisualEhrmanntraut** [comment]
> > @waterdragon78 One of the requirements to get to the UI is getting past the data migration step. The data migration step is ran even if the device has been newly restored, and it installs things like apps (Calculator, etc).
> > We have attempted to patch all the blocking requirements out and the UI does show up and if we bypass the activation we also get to the Home Screen. Below is how that looks like:
> > Screenshots
> 
> Hey, I'm super curious what patches you made to get the UI to show up! I've been trying to get past the same Data Migration step on iOS 15. I've been working on hacking iOS to run on vma2 (Virtualization.framework), and have only gotten PreBoard.app to show up, otherwise, my attempts to patch Data Migration seem to get stuck on a black screen, and I wonder if your patches would make the difference...

Responded about this to your Telegram message.

**Comment by VisualEhrmanntraut** [comment]
By the way, I should mention that me and mostly @chris-pcguy has silently been working on getting sepOS emulation working the past year. He seems to be reaching close to the finish line. It is possible we'll get the emulated SEP working before the simulated one.

**Comment by Anonymous941** [comment]
If this is completed, where will you put it?  Will there be a comment on this issue or somewhere else?

**Comment by VisualEhrmanntraut** [comment]
I forgot to mention that SEP emulation was indeed completed, you can see it in the feat-sep_emu branch

**Comment by Anonymous941** [comment]
@VisualEhrmanntraut Does this mean I can try to set it up now, or is there another hurdle?

**Comment by VisualEhrmanntraut** [comment]
Well, no, there's still a lot of hoops to jump through. You need a decrypted SEP/OS, SEP ROM, you need to patch the filesystem to remove some graphics-acceleration-related things.

**Comment by VisualEhrmanntraut** [comment]
(The filesystem patches will be released soon-ish, I'm waiting on the security research company that made these patches to see whether I'll be releasing them or if they'll be. But I will let everyone know here.)

**Comment by VisualEhrmanntraut** [comment]
(amongst the patches will be a data migration skip pathc, so you can use the incomplete SEP simulation in case anyone wants to avoid the whole SEP emulation)

**Comment by jessem542** [comment]
> (amongst the patches will be a data migration skip patch, so you can use the incomplete SEP simulation in case anyone wants to avoid the whole SEP emulation)

So, simulated SEP requires to skip data migration step and that's why most default apps won't show up in the screenrecord video you shared? With a full emulated SEP, those apps could be loaded into the system and we could expect a normal iOS behavior (beside graphics acceleration, bugs and cpu slow emulation)?

**Comment by VisualEhrmanntraut** [comment]
@jessem542 I guess.

**Comment by agnosticlines** [comment]
> (The filesystem patches will be released soon-ish, I'm waiting on the security research company that made these patches to see whether I'll be releasing them or if they'll be. But I will let everyone know here.)

Hey, have eShard said if they're committing to open sourcing the patches/work?

**Comment by fuad00** [comment]
> > (The filesystem patches will be released soon-ish, I'm waiting on the security research company that made these patches to see whether I'll be releasing them or if they'll be. But I will let everyone know here.)
> 
> Hey, have eShard said if they're committing to open sourcing the patches/work?

are you talking about them? https://youtube.com/watch?v=2P215fXC6JY

looks like they’ve build docker compose … 🫨

**Comment by jessem542** [comment]
> > (The filesystem patches will be released soon-ish, I'm waiting on the security research company that made these patches to see whether I'll be releasing them or if they'll be. But I will let everyone know here.)
> 
> Hey, have eShard said if they're committing to open sourcing the patches/work?

They said now are evaluating legal related things before launching. They seems very open to release it, if they can't I hope @VisualEhrmanntraut could do it. Just we shouldn't rush things.

**Comment by VisualEhrmanntraut** [comment]
I am working with some people on creating a minimal or no patch setup, of which its instructions will be added to a wiki in this repo.

**Comment by agnosticlines** [comment]
> > > (The filesystem patches will be released soon-ish, I'm waiting on the security research company that made these patches to see whether I'll be releasing them or if they'll be. But I will let everyone know here.)
> > 
> > 
> > Hey, have eShard said if they're committing to open sourcing the patches/work?
> 
> are you talking about them? https://youtube.com/watch?v=2P215fXC6JY
> 
> looks like they’ve built docker compose … 🫨

Docker compose is just the mechanism they use to deploy esREVEN and other parts of their infrastructure, they have done some great work getting iOS 15 booting, writing the patches etc, let's not downplay the work they've done



> > > (The filesystem patches will be released soon-ish, I'm waiting on the security research company that made these patches to see whether I'll be releasing them or if they'll be. But I will let everyone know here.)
> > 
> > 
> > Hey, have eShard said if they're committing to open sourcing the patches/work?
> 
> They said now are evaluating legal related things before launching. They seems very open to release it, if they can't I hope [@VisualEhrmanntraut](https://github.com/VisualEhrmanntraut) could do it. Just we shouldn't rush things.

Makes sense! Thanks :-)

**Comment by VisualEhrmanntraut** [comment]
> Docker compose is just the mechanism they use to deploy esREVEN and other parts of their infrastructure, they have done some great work getting iOS 15 booting, writing the patches etc, let's not downplay the work they've done

eShard only made filesystem patches in order to bypass the missing features, so that we can figure out what exactly is going wrong. The emulator development has been done solely by me and my colleague Christian. We tried to warn the company's marketing team that the phrasing is slightly misleading but they ignored the founder's instructions 🤷‍♂️

**Comment by agnosticlines** [comment]
> > Docker compose is just the mechanism they use to deploy esREVEN and other parts of their infrastructure, they have done some great work getting iOS 15 booting, writing the patches etc, let's not downplay the work they've done
> 
> eShard only made filesystem patches in order to bypass the missing features, so that we can figure out what exactly is going wrong. The emulator development has been done solely by me and my colleague Christian. We tried to warn the company's marketing team that the phrasing is slightly misleading but they ignored the founder's instructions 🤷‍♂️

Oh I thought they made more changes than just the patches :-( My bad. I thought your fork didn't have a graphics stack/framebuffer implemented yet for example

**Comment by VisualEhrmanntraut** [comment]
In other words, all the changes are already pushed, and with the emulated SEP branch in combination with the filesystem patches, the UI shows up.

**Comment by agnosticlines** [comment]
> The display (framebuffer) was implemented by me, that's why you can see an Apple Logo. The GPU isn't, and part is bypassed through those aforementioned filesystem patches.

Ah, so the filesystem patches skip the need for the GPU (I guess also leading to those visual artifacts like square icons?), that's interesting! I knew that the apple logo booting didn't need the full GPU just a framebuffer, I'm not 100% sure why I said "/framebuffer", that's on me

I really appreciate all your work, out of curiosity do you know how much work it'd be to go from iOS 15 to iOS 18? Trungs QEMU fork was targeting iPhone 11 IIRC so no major hardware changes needed there if I'm understanding correctly?

**Comment by VisualEhrmanntraut** [comment]
@agnosticlines I forgot to mention this is iOS 14 beta 5, not iOS 15. iOS 14 beta 5 is the only version that works right now, any newer versions will need the code to be refactored so that the addresses aren’t hardcoded. Hardware itself doesn’t need changes most likely.

**Comment by infinite9898** [comment]
Guys can anyone help me how can I get this running on my linux. I am using a NVIDIA GPU, does it support running it?

**Comment by VisualEhrmanntraut** [comment]
Hello everyone. I wanted to give you an update so you don't think we ran away or got silenced. We're trying to replicate the results presented here but we are failing miserably. All of a sudden things aren't working anymore. We're trying hard to replicate the setup in the background.

**Comment by jessem542** [comment]
> Hello everyone. I wanted to give you an update so you don't think we ran away or got silenced. We're trying to replicate the results presented here but we are failing miserably. All of a sudden things aren't working anymore. We're trying hard to replicate the setup in the background.

Take your time guys, we are not rushing things. May there's something we can help?

**Comment by VisualEhrmanntraut** [comment]
Unfortunately I don't think there's much you can do. We had a set of patches that should've been working but for some reason don't anymore. Hopefully we figure it out eventually.

**Comment by jessem542** [comment]
> Unfortunately I don't think there's much you can do. We had a set of patches that should've been working but for some reason don't anymore. Hopefully we figure it out eventually.

Hmmm, so what's exactly the problem? a kernel panic? Or just the system suddenly can't boot? What the QEMU logs says?
You guys are still working with the company? If so, they got the same problem?

**Comment by VisualEhrmanntraut** [comment]
The problem is that the UI won't display anymore (data migration, then just stuck on apple logo or black screen). Nothing useful in logs as far as we can tell. And for this specific task (creating the minimal set of changes to create a working setup) that company is not involved, we just want to provide you all with a temporary way to play around with the emulator.

**Comment by jessem542** [comment]
> The problem is that the UI won't display anymore (data migration, then just stuck on apple logo or black screen). Nothing useful in logs as far as we can tell. And for this specific task (creating the minimal set of changes to create a working setup) that company is not involved, we just want to provide you all with a temporary way to play around with the emulator.

Interesting, this trouble started within this minimal setup emulator you guys are preparing or also on full patched emulator? It looks like you guys need to debug the latest changes you applied but I'm not familiar with what kind of debug tool you can use in for this case. Wishing you the best luck on this!

**Comment by VisualEhrmanntraut** [comment]
I meant the previous setup doesn't work either. And the latest changes came as a result of debugging. We're still trying to figure out what exactly is the issue.

**Comment by jessem542** [comment]
> I meant the previous setup doesn't work either. And the latest changes came as a result of debugging. We're still trying to figure out what exactly is the issue.

You guys got this, wishing you the best luck! 🐐

**Comment by VisualEhrmanntraut** [comment]
we got the UI to show up once but subsequent runs seem to just do this
https://github.com/user-attachments/assets/46325282-1418-4abf-84ab-761e1672e65c

**Comment by jessem542** [comment]
> we got the UI to show up once but subsequent runs seem to just do this
> 
> https://github.com/user-attachments/assets/46325282-1418-4abf-84ab-761e1672e65c
> 

I thought you guys was getting springboard through VNC, this seems like framebuffer display implemented. Is that right? Either way this could relate to graphics acceleration issue? May iOS is still trying to render qe/ci despite the already injected patches.

**Comment by VisualEhrmanntraut** [comment]
It was QEMU's screen through VNC, like I said the framebuffer is implemented and working.

**Comment by jessem542** [comment]
> It was QEMU's screen through VNC, like I said the framebuffer is implemented and working.

So, what do you think it could be the problem?

**Comment by agnosticlines** [comment]
> we got the UI to show up once but subsequent runs seem to just do this
> 
>  Screen.Recording.2024-12-14.at.17.47.16.mp4

When you say "once" is that once repeatably? Like if you have a fresh system it'll show the UI once but after a reboot it no longer will, or is that you can't get it to show the UI at all anymore

**Comment by VisualEhrmanntraut** [comment]
Looks like this now, by the way:

https://github.com/user-attachments/assets/09254fc6-f019-4492-a9e0-e32960fbb9e9

**Comment by jessem542** [comment]
> Looks like this now, by the way:
> 
> https://github.com/user-attachments/assets/09254fc6-f019-4492-a9e0-e32960fbb9e9
> 

Damn guys this is sick! I suppose is not possible enable software rendering for the patches, right? Something like Corellium did.

**Comment by jessem542** [comment]
> Looks like this now, by the way:
> 
> https://github.com/user-attachments/assets/09254fc6-f019-4492-a9e0-e32960fbb9e9
> 

Damn guys this is sick! I suppose is not possible enable software rendering for the patches, right? Something like Corellium did.

**Comment by VisualEhrmanntraut** [comment]
This is software rendering, using a few patches. Corellium's platform has some basic graphics acceleration.

**Comment by VisualEhrmanntraut** [comment]
So, it seems like the data migration might stuck due to backboardd waiting for the screen digitiser. So it affects the SEP emulation as well.

**Comment by VisualEhrmanntraut** [comment]
As in, data migration works with the SEP emulation, but only with the UI-related patches applied, one of them being disabling the function that waits for the screen digitiser, so it’s safe to assume the problem is likely that.

**Comment by VisualEhrmanntraut** [comment]
An update for you guys:

The two blocking issues right now are the following:
- Missing Multitouch
- Unhappy CommCenter (probably due to missing bluetooth/wifi and baseband)

**Comment by VisualEhrmanntraut** [comment]
2nd can be bypassed by disabling CommCenter, first one with a patch and a VNC server.

I am responsible for implementing multitouch. This is my progress:

[T8030Stderr.txt](https://github.com/user-attachments/files/18389378/T8030Stderr.txt)
[T8030Stdout.txt](https://github.com/user-attachments/files/18389379/T8030Stdout.txt)

**Comment by jessem542** [comment]
> An update for you guys:
> 
> The two blocking issues right now are the following:
> - Missing Multitouch
> - Unhappy CommCenter (probably due to missing bluetooth/wifi and baseband)

This is so interesting, enabling multitouch it will enable using the system without patches and VNC then. I'm wondering how hard will be writing drivers to fake wifi/Bluetooth as dummy hardware. Something like Corellium devices has implemented. So that's why disabling CommCenter is a go to go. 

My question is, what could be the best practice to enable Internet within the VM?
I'd love to contribute to this project once is released, good work!

**Comment by VisualEhrmanntraut** [comment]
> My question is, what could be the best practice to enable Internet within the VM?

Without WiFi emulation, a reverse proxy could work I assume.

**Comment by VisualEhrmanntraut** [comment]
> This is so interesting, enabling multitouch it will enable using the system without patches and VNC then.

Not exactly, still need to modify the filesystem to at least disable GPU acceleration for some frameworks. Not sure if there's any workaround, we're still working on it. But we're working very hard.

**Comment by jessem542** [comment]
######

**Comment by jessem542** [comment]
##########

**Comment by itsnotrin** [comment]
Is there anything that the wider community can do to help with the project?

**Comment by itsnotrin** [comment]
I meant more in terms of Development help but that works too 😂

**Comment by itsnotrin** [comment]
If you get at least basic build docs live I'm happy to try and work on anything needed- understand if not and it's more of a "figure it out yourself" kind of story though.

**Comment by jeppojeps** [comment]
Hello guys, thanks a lot for your huge contribution, I have been looking at your project very close and collaborated with Martjin in some tests for the ipod_touch 2g which sometimes I have used in my classes, I had tried to make the shell working but there was some major functionality working i.e. no network and no usb and that blocked my port of the aleph research tunnel to use usbmuxd. 
Visual was mentioning multitouch and I remember that Martjin implemented it for the ipod touch 2g
https://github.com/devos50/qemu-ios/blob/ipod_touch_2g/hw/arm/ipod_touch_multitouch.c

I am not sure how useful it can be on a complex device such as the ip11. But maybe would serve as an inspiration. 

Cheers.

**Comment by VisualEhrmanntraut** [comment]
Hello there. I am aware of this code base, the code for the multitouch was a bit useful but unfortunately we are unlucky and they changed the multitouch firmware communication exactly on the t8030 chip, every chip before that used the same communication as the code you’ve linked.

**Comment by jeppojeps** [comment]
<sigh> thanks for the quick response...

**Comment by VisualEhrmanntraut** [comment]
I got it working on my machine yesterday, so now I know exactly what to do:

https://github.com/user-attachments/assets/f752848d-d0c0-44fb-b5b3-b43fb00a28fd

**Comment by goodtomeetyou** [comment]
> I got it working on my machine yesterday, so now I know exactly what to do about the data migration:
> 
> https://github.com/user-attachments/assets/f752848d-d0c0-44fb-b5b3-b43fb00a28fd

wow, so data migration struggles are not a problem now? 😮

**Comment by VisualEhrmanntraut** [comment]
Technically not. I know the exact problems and workarounds, and can come up with solutions.

**Comment by goodtomeetyou** [comment]
> I got it working on my machine yesterday, so now I know exactly what to do about the data migration:
> 
> https://github.com/user-attachments/assets/f752848d-d0c0-44fb-b5b3-b43fb00a28fd

btw, I see you're working on Apple Sillicon, it's KVM virtualization or just arm64 QEMU emulation? can't wait to try Aleph research virtualization patches

**Comment by VisualEhrmanntraut** [comment]
aleph security doesn't have anything to do with me lol. and no it's not kvm.

**Comment by goodtomeetyou** [comment]
> Lol, this is ChefKiss, not Aleph Security, we are not related to that team but I have talked with Afek from them. And it's emulated, not KVM (and KVM is a Linux thing, macOS has Hypervisor.framework).

Yes I am aware of that. Ofc I know this is not that team. I just mentioned because I am curious how those virtualization patches could work on this xnu-qemu fork once it's ready and try them on Linux since it's Unix based system. I think I didn't expressed well so sorry for that confusion tho 🙃

Anyways, I'll keep following your progress closely. What's your new target once data migration will be fixed?

**Comment by VisualEhrmanntraut** [comment]
https://github.com/user-attachments/assets/a4aeef3c-a53d-4fa3-8ec9-130db44ad7d6

**Comment by VisualEhrmanntraut** [comment]
iOS 17 is now partially booting

<img width="1624" alt="Image" src="https://github.com/user-attachments/assets/69e58675-9eb9-41ff-9f34-bf89c38e9e17" />

**Comment by goodtomeetyou** [comment]
> iOS 17 is now partially booting
> 
Wow, such an awesome progress. There's a kernel panic due to memory mapping differences, how did you used to handle SMC on iOS 14? 
You guys are considering add support for newer iOS versions or you're just testing?

**Comment by VisualEhrmanntraut** [comment]
> ou guys are considering add support for newer iOS versions or you're just testing?

the former

**Comment by Anonymous941** [comment]
Do you think once you finish these 3 things it will be worth it for me to try installing it?  As far as I can tell, currently it requires patches that are complicated to apply properly and require macOS, because otherwise the OS will crash (even though they're not needed).  Once they're done, will the patches no longer be necessary?

**Comment by chris-pcguy** [comment]
> Do you think once you finish these 3 things

Feel free to implement gpu acceleration yourself.

> it will be worth it for me to try installing it?

Nobody is forcing you to use it.

> As far as I can tell, currently it requires patches that are complicated to apply properly

Not complicated at all, there are quite a few approaches to script that.

> and require macOS

For now, until read-write apfs on e.g. Linux, BSD or Windows gets more reliable.

> because otherwise the OS will crash (even though they're not needed)

Crashing, as in kernel panic? It won't crash without the user-space patches.

> Once they're done, will the patches no longer be necessary?

The plan is to make it require as few patches as possible.

**Comment by jeppojeps** [comment]
Thanks for your response, @chris-pcguy . I feel there’s a lack of understanding about the immense effort required to make this work properly. Some questions can come off as disrespectful, even if unintentionally.

Exposing projects to the public comes with challenges, but it also has immense educational value, even when unfinished. A great example is the iPod2g, which provides invaluable insights. Many seem to overlook Moore’s law and the exponential complexity of modern devices. While emulating older systems like the Game Boy, PS1, or iPod2g is challenging but feasible, emulating a currently produced device is a completely different level of complexity.

Let’s respect the work being done here. It’s a monumental task, and the dedication behind it deserves appreciation.

**Comment by Anonymous941** [comment]
> Let’s respect the work being done here. It’s a monumental task, and the dedication behind it deserves appreciation.

Sorry if I implied otherwise, I think there's a misunderstanding and I'm sorry if I phrased my comment wrong.  I am amazed by this finally existing (and hopefully my earlier comments express that), after so many years of only one proprietary cloud service and a bunch of scams

> Nobody is forcing you to use it.

This is a misunderstanding - I wasn't trying to imply that the process was unnecessarily complicated or that it's not worth my time.  I mistakenly thought the patches were temporary and would be depreciated soon, and didn't want to have to ask a bunch of questions to trying to get them working

**Comment by VisualEhrmanntraut** [comment]
In other words, the current situation is:
- Software rendering for some parts can be activated via environment variable. This mostly fixes the data migration as most of the crashing processes are no longer crashing.
- Other parts are activated automatically when there is no Metal GPU present.
- Some functions, at least in iOS 14.0 Beta 5, are assuming that there is a GPU present, and access a null pointer causing a crash. This causes the actual UI, SpringBoard and potentially its data migrators, to crash. Dyld Shared Cache patches have been manually applied to remove the faulty code.
- As this is iOS, and the device is an iPhone, many places are assuming Baseband should be here, and are waiting for it. We have worked around that by disabling some of the launch services. This one will be able to be fixed after Apple PCI-e and Baseband is implemented.

**Comment by VisualEhrmanntraut** [comment]
@Anonymous941 Does this help you understand better?

**Comment by VisualEhrmanntraut** [comment]
Oh, and not all patches can be removed, at least not kernel patches, as they are required for iOS versions that are not longer signed.

**Comment by Anonymous941** [comment]
> [@Anonymous941](https://github.com/Anonymous941) Does this help you understand better?

Yes, thanks for the explaination.  It doesn't really matter that much to me personally, I can always just make a macOS VM to apply them, or dig out my old MacBook.  I was assuming that it would require unpacking some obscure proprietary format (on second thought, not sure why I assumed that).  Can jailbroke iOS have RW access to APFS images by any chance?
> Oh, and not all patches can be removed, at least not kernel patches, as they are required for iOS versions that are not longer signed.

Can't you import the keys from a real iPhone with SHSH2 blobs?  Or would that require more hardware implemented?  (of course it's moot atm because some of the patches are still needed)
> Maybe for Linux hosts, but I don't use Linux, so I can't test or implement it.
This type of low-level programming is outside of my skill set, but I'd be happy to assist with testing if anyone ever wants to implement this in the future

**Comment by VisualEhrmanntraut** [comment]
> I was assuming that it would require unpacking some obscure proprietary format (on second thought, not sure why I assumed that).

Nah. The ipsw is just a zip, and the NAND is more or less just standard NVMe with a few namespaces and in one of them being the main filesystem, which is just GPT partitioning and an APFS partition with data/system/etc volumes like macOS.

> Can jailbroke iOS have RW access to APFS images by any chance?

Don't know.

> Can't you import the keys from a real iPhone with SHSH2 blobs?

The problem is that if you don't have the blobs, the files aren't signed, so the kernel rejects some things e.g. the touch screen firmware (so the touch screen stops working). For this particular iOS version, we don't have them, so we have to forge the ticket, so the signatures aren't really valid.

**Comment by goodtomeetyou** [comment]
> This is a complicated task on its own, but with QEMU probably not even possible to implement at least for macOS hosts due to OpenGL support not being able to be enabled.

This is a research area very complex. I'm also researching how implement AGX emulation on ARM64 macOS. May we could start with MoltenVK. This is why you want to start your own emulator from scratch? Due to lack of QEMU capabilities for graphics emulation?

**Comment by VisualEhrmanntraut** [comment]
<details>
<summary>iOS 15</summary>

![Image](https://github.com/user-attachments/assets/50efe557-c3a3-4f40-947a-dbea345a8274)

![Image](https://github.com/user-attachments/assets/a23f5d0c-65ba-4ccf-8e00-7b3c849e326c)

</details>

<details>
<summary>iOS 16</summary>

![Image](https://github.com/user-attachments/assets/6b176ce0-a379-42eb-8974-e5a4252ae983)

![Image](https://github.com/user-attachments/assets/b31ca528-58e7-4948-b172-5f0e10dcef74)

![Image](https://github.com/user-attachments/assets/d4c51507-433b-47ca-8625-a0fe0df558ce)

</details>

**Comment by goodtomeetyou** [comment]
> iOS 15
> ![Image](https://github.com/user-attachments/assets/50efe557-c3a3-4f40-947a-dbea345a8274)
> 
> ![Image](https://github.com/user-attachments/assets/a23f5d0c-65ba-4ccf-8e00-7b3c849e326c)
> 
> iOS 16
> ![Image](https://github.com/user-attachments/assets/6b176ce0-a379-42eb-8974-e5a4252ae983)
> 
> ![Image](https://github.com/user-attachments/assets/b31ca528-58e7-4948-b172-5f0e10dcef74)
> 
> ![Image](https://github.com/user-attachments/assets/d4c51507-433b-47ca-8625-a0fe0df558ce)
Crazy af! I've been seeing the commits lately, seems like the emulation for multiple iOS versions it's getting pretty close. I am curious, what's your next goal for the project these upcoming months?

**Comment by GradedWarrior** [comment]
Saw this today https://youtu.be/yuOcZyeUyN0

**Comment by goodtomeetyou** [comment]
> https://github.com/ChefKissInc/QEMUAppleSilicon/wiki

You a goat bro, thanks for your hard work you did here with Chris. You will continue researching and developing this project (audio, AMX, etc.) or you'll move on with other projects?

**Comment by waterdragon78** [comment]
So happy to see this, the passion and work put into this project is remarkable! Thank yall!

**Comment by waterdragon78** [comment]
<img width="436" alt="Image" src="https://github.com/user-attachments/assets/3c1f015b-bf2c-4064-9eef-ab053ba3c883" />

Holy crap this is like coolest thing ever! I know I already said it, but serious props to you guys. Never in a million years would I have thought something of this caliber would be possible. Super excited to see what's to come!
